api/src/main/java/org/apache/cloudstack/api/command/admin/systemvm/MigrateSystemVMCmd.java

```diff
@@ -122,34 +122,34 @@ public String getEventDescription() {
-            throw new InvalidParameterValueException("Either hostId or storageId must be specified");
-        }
-
         if (getHostId() != null && getStorageId() != null) {
             throw new InvalidParameterValueException("Only one of hostId and storageId can be specified");
         }
+
         try {
             //FIXME : Should not be calling UserVmService to migrate all types of VMs - need a generic VM layer
             VirtualMachine migratedVm = null;
-            if (getHostId() != null) {
-                Host destinationHost = _resourceService.getHost(getHostId());
-                if (destinationHost == null) {
-                    throw new InvalidParameterValueException("Unable to find the host to migrate the VM, host id=" + getHostId());
-                }
-                if (destinationHost.getType() != Host.Type.Routing) {
-                    throw new InvalidParameterValueException("The specified host(" + destinationHost.getName() + ") is not suitable to migrate the VM, please specify another one");
-                }
-                CallContext.current().setEventDetails("VM Id: " + getVirtualMachineId() + " to host Id: " + getHostId());
-                migratedVm = _userVmService.migrateVirtualMachineWithVolume(getVirtualMachineId(), destinationHost, new HashMap<String, String>());
-            } else if (getStorageId() != null) {
+            if (getStorageId() != null) {
                 // OfflineMigration performed when this parameter is specified
                 StoragePool destStoragePool = _storageService.getStoragePool(getStorageId());
                 if (destStoragePool == null) {
                     throw new InvalidParameterValueException("Unable to find the storage pool to migrate the VM");
                 }
                 CallContext.current().setEventDetails("VM Id: " + getVirtualMachineId() + " to storage pool Id: " + getStorageId());
                 migratedVm = _userVmService.vmStorageMigration(getVirtualMachineId(), destStoragePool);
+            } else {
+                Host destinationHost = null;
+                if (getHostId() != null) {
+                    destinationHost =_resourceService.getHost(getHostId());
+                    if (destinationHost == null) {
+                        throw new InvalidParameterValueException("Unable to find the host to migrate the VM, host id=" + getHostId());
+                    }
+                    if (destinationHost.getType() != Host.Type.Routing) {
+                        throw new InvalidParameterValueException("The specified host(" + destinationHost.getName() + ") is not suitable to migrate the VM, please specify another one");
+                    }
+                }
+                CallContext.current().setEventDetails("VM Id: " + getVirtualMachineId() + " to host Id: " + getHostId());
+                migratedVm = _userVmService.migrateVirtualMachineWithVolume(getVirtualMachineId(), destinationHost, new HashMap<String, String>());
```

Comment:
@weizhouapache change added in UserVmManagerImpl to find suitable host is only for `migrateVirtualMachine` method while we call `migrateVirtualMachineWithVolume`
Currently we will get NPE onhttps://github.com/apache/cloudstack/blob/f5aed32ebddef19b9b2f1bed724bf0b16c69f0b6/server/src/main/java/com/cloud/vm/UserVmManagerImpl.java#L6289

Reply:
@shwstppr I will test it.

Reply:
@shwstppr I tested with 4.15. did not notice that we have some changes in master.
https://github.com/apache/cloudstack/blob/4.15/api/src/main/java/org/apache/cloudstack/api/command/admin/systemvm/MigrateSystemVMCmd.java#L120

Reply:
@shwstppr I have a question to you, since you are the author of #4385.
why we use different method in vm migration and systemvm migration ?

MigrateSystemVMCmd.java: migrateVirtualMachineWithVolume and vmStorageMigration
MigrateVMCmd.java: migrateVirtualMachine and vmStorageMigration

can they use same methods ?

Reply:
@weizhouapache For user VMs we have two different APIs, migrateVirtualMachine and migrateVirtualMachineWithVolume. Depending on the need for storage migration requirement appropriate API can be used.
While working on #4385 choice was to add a new API migrateSystemVmWithVolume or using `migrateVirtualMachineWithVolume` method while code making the decision if volume needs to be migrated.

Reply:
@shwstppr we pass a empty hashmap in `migrateVirtualMachineWithVolume `, see below
https://github.com/apache/cloudstack/blob/master/api/src/main/java/org/apache/cloudstack/api/command/admin/systemvm/MigrateSystemVMCmd.java#L144

so systemvm migration actually works without volume ?
I changed it back to 4.15, systemvm migration works
https://github.com/apache/cloudstack/blob/4.15/api/src/main/java/org/apache/cloudstack/api/command/admin/systemvm/MigrateSystemVMCmd.java#L120

Reply:
@weizhouapache yes, we do pass empty volume-pool mapping to the method but later in the code, we  generate volume-pool mapping when needed (such as inter-cluster migration with cluster-scoped pools)
Moving back to 4.15 code will definitely allow systemvm migration with changes from this PR but inter-cluster migration of systemvms would not work.

Reply:
@shwstppr I was a bit confused.

I will change to the following behavior: if hostid is not specified, then use migratevirtualmachine instead of migratevirtualmachinewithvolume. sounds ok ?

Reply:
@weizhouapache Yes, works fine in that case. Tested,

```
(localcloud) SBCM5> > migrate systemvm virtualmachineid=c3737243-93b2-4369-bf60-08ac217e3465
{
  "systemvm": {
    "agentstate": "Up",
    "created": "2021-03-16T11:22:18+0000",
    "dns1": "10.0.32.1",
    "dns2": "8.8.8.8",
    "gateway": "10.0.48.1",
    "hostid": "2e5b053f-539e-45c7-80f3-0fe3f0291c05",
    "hostname": "10.0.33.141",
    "hypervisor": "VMware",
    "id": "c3737243-93b2-4369-bf60-08ac217e3465",
    "linklocalmacaddress": "02:00:5e:cd:00:01",
    "name": "s-1-VM",
    "podid": "57382744-92f2-4610-a873-733de71eb142",
    "podname": "Pod1",
    "privateip": "10.0.36.123",
    "privatemacaddress": "1e:00:47:00:00:17",
    "privatenetmask": "255.255.240.0",
    "publicip": "10.0.52.121",
    "publicmacaddress": "1e:00:70:00:00:01",
    "publicnetmask": "255.255.240.0",
    "state": "Running",
    "systemvmtype": "secondarystoragevm",
    "templateid": "ff29ad45-6b1d-4c8e-aa6e-74f41b483134",
    "templatename": "SystemVM Template (vSphere)",
    "version": "4.16.0.0-SNAPSHOT",
    "zoneid": "8f79f870-abc3-4d46-bb0e-cdb695ca45cd",
    "zonename": "pr4378-t146-vmware-67u3"
  }
}
```